shutdown_test.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx
+
+import (
+	"context"
+	"syscall"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestShutdown(t *testing.T) {
+	t.Run("ShutsDownApp", func(t *testing.T) {
+		var started, stopped bool
+		app := New(Invoke(func(l Lifecycle, s Shutdowner) {
+			l.Append(Hook{
+				OnStart: func(ctx context.Context) error {
+					started = true
+					return nil
+				},
+				OnStop: func(ctx context.Context) error {
+					stopped = true
+					return nil
+				},
+			})
+			l.Append(Hook{
+				OnStart: func(ctx context.Context) error { return s.Shutdown() },
+			})
+		}))
+
+		app.Run()
+		assert.True(t, started, "app wasn't started")
+		assert.True(t, stopped, "app wasn't stopped")
+	})
+
+	t.Run("BroadcastsToMultipleChannels", func(t *testing.T) {
+		var started, stopped bool
+		app := New(Invoke(func(l Lifecycle, s Shutdowner) {
+			l.Append(Hook{
+				OnStart: func(ctx context.Context) error {
+					started = true
+					return nil
+				},
+				OnStop: func(ctx context.Context) error {
+					stopped = true
+					return nil
+				},
+			})
+		}))
```

Comment:
We've got this pattern in a few places now. Can we factor it out into a reusable option?

```
type lifecycleVerifier struct {
  fx.Option

  started, stopped bool
}

func newLifecycleVerifier() *lifecycleVerifier {
  var started, stopped bool
  opt := Invoke(func(l Lifecycle) {
    l.Append(Hook{
      OnStart: func(context.Context) error {
        started = true
        return nil
      },
      OnStop: func(context.Context) error {
        stopped = true
        return nil
      },
    })
  })
  return &lifecycleVerifier{
    Option: opt,
    started: started,
    stopped: stopped,
  }
}
```

For bonus points, have the verifier struct include a `testing.TB` and add methods that assert that `started` and `stopped` are true.

(There are probably some errors in there, but hopefully you get the point.)